[Bug] Version 2.1 Submitting a batch task is abnormal

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-streampark/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### Java Version

jdk 1.8

### Scala Version

2.11.x

### StreamPark Version

2.1.0

### Flink Version

1.14.6

### deploy mode

yarn-application

### What happened

An exception occurred when the sql was submitted in batch mode


### Error Exception

```log
2023-08-25 11:42:33,149 WARN  org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap [] - Application failed unexpectedly:
java.util.concurrent.CompletionException: org.apache.flink.client.deployment.application.ApplicationExecutionException: Could not execute application.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:943) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[?:1.8.0_181]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:287) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.lambda$runApplicationAsync$2(ApplicationDispatcherBootstrap.java:224) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_181]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_181]
	at org.apache.flink.runtime.concurrent.akka.ActorSystemScheduledExecutorAdapter$ScheduledFutureTask.run(ActorSystemScheduledExecutorAdapter.java:171) ~[flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68) ~[flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.lambda$withContextClassLoader$0(ClassLoadingUtils.java:41) ~[flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49) [flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48) [flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) [?:1.8.0_181]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) [?:1.8.0_181]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) [?:1.8.0_181]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) [?:1.8.0_181]
Caused by: org.apache.flink.client.deployment.application.ApplicationExecutionException: Could not execute application.
	... 13 more
Caused by: org.apache.flink.client.program.ProgramInvocationException: The main method caused an error: No operators defined in streaming topology. Cannot execute.
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:372) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:114) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:261) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	... 12 more
Caused by: java.lang.IllegalStateException: No operators defined in streaming topology. Cannot execute.
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.getStreamGraphGenerator(StreamExecutionEnvironment.java:2105) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.generateStreamGraph(StreamExecutionEnvironment.java:2100) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.table.planner.delegation.DefaultExecutor.createPipeline(DefaultExecutor.java:83) ~[flink-table_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.execute(TableEnvironmentImpl.java:1606) ~[flink-table_2.11-1.14.6.jar:1.14.6]
	at org.apache.streampark.flink.core.TableContext.execute(TableContext.scala:68) ~[streampark-flink-shims_flink-1.14_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.core.FlinkTableTrait.start(FlinkTableTrait.scala:39) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.core.scala.FlinkTable$class.main(FlinkTable.scala:42) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient$BatchSqlApp$.main(SqlClient.scala:64) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient$.delayedEndpoint$org$apache$streampark$flink$cli$SqlClient$1(SqlClient.scala:57) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient$delayedInit$body.apply(SqlClient.scala:31) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at scala.Function0$class.apply$mcV$sp(Function0.scala:34) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.App$$anonfun$main$1.apply(App.scala:76) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.App$$anonfun$main$1.apply(App.scala:76) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.collection.immutable.List.foreach(List.scala:392) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.App$class.main(App.scala:76) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.streampark.flink.cli.SqlClient$.main(SqlClient.scala:31) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient.main(SqlClient.scala) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_181]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_181]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_181]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_181]
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:114) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:261) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	... 12 more
2023-08-25 11:42:33,155 INFO  org.apache.flink.runtime.resourcemanager.active.ActiveResourceManager [] - Requesting new worker with resource spec WorkerResourceSpec {cpuCores=30.0, taskHeapSize=15.292gb (16419651555 bytes), taskOffHeapSize=125.000mb (131072000 bytes), networkMemSize=1024.000mb (1073741824 bytes), managedMemSize=1.852gb (1988100125 bytes), numSlots=30}, current pending count: 1.
2023-08-25 11:42:33,156 ERROR org.apache.flink.runtime.entrypoint.ClusterEntrypoint        [] - Fatal error occurred in the cluster entrypoint.
java.util.concurrent.CompletionException: org.apache.flink.client.deployment.application.ApplicationExecutionException: Could not execute application.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:943) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:1.8.0_181]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[?:1.8.0_181]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:287) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.lambda$runApplicationAsync$2(ApplicationDispatcherBootstrap.java:224) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_181]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_181]
	at org.apache.flink.runtime.concurrent.akka.ActorSystemScheduledExecutorAdapter$ScheduledFutureTask.run(ActorSystemScheduledExecutorAdapter.java:171) ~[flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68) ~[flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.lambda$withContextClassLoader$0(ClassLoadingUtils.java:41) ~[flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49) [flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48) [flink-rpc-akka_16be5c59-5ecf-4c7e-808f-900a6875ceae.jar:1.14.6]
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289) [?:1.8.0_181]
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056) [?:1.8.0_181]
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692) [?:1.8.0_181]
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157) [?:1.8.0_181]
Caused by: org.apache.flink.client.deployment.application.ApplicationExecutionException: Could not execute application.
	... 13 more
Caused by: org.apache.flink.client.program.ProgramInvocationException: The main method caused an error: No operators defined in streaming topology. Cannot execute.
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:372) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:114) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:261) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	... 12 more
Caused by: java.lang.IllegalStateException: No operators defined in streaming topology. Cannot execute.
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.getStreamGraphGenerator(StreamExecutionEnvironment.java:2105) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.generateStreamGraph(StreamExecutionEnvironment.java:2100) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.table.planner.delegation.DefaultExecutor.createPipeline(DefaultExecutor.java:83) ~[flink-table_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.execute(TableEnvironmentImpl.java:1606) ~[flink-table_2.11-1.14.6.jar:1.14.6]
	at org.apache.streampark.flink.core.TableContext.execute(TableContext.scala:68) ~[streampark-flink-shims_flink-1.14_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.core.FlinkTableTrait.start(FlinkTableTrait.scala:39) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.core.scala.FlinkTable$class.main(FlinkTable.scala:42) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient$BatchSqlApp$.main(SqlClient.scala:64) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient$.delayedEndpoint$org$apache$streampark$flink$cli$SqlClient$1(SqlClient.scala:57) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient$delayedInit$body.apply(SqlClient.scala:31) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at scala.Function0$class.apply$mcV$sp(Function0.scala:34) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.App$$anonfun$main$1.apply(App.scala:76) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.App$$anonfun$main$1.apply(App.scala:76) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.collection.immutable.List.foreach(List.scala:392) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at scala.App$class.main(App.scala:76) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.streampark.flink.cli.SqlClient$.main(SqlClient.scala:31) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at org.apache.streampark.flink.cli.SqlClient.main(SqlClient.scala) ~[streampark-flink-sqlclient_2.11-2.1.0.jar:2.1.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_181]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_181]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_181]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_181]
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:114) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	at org.apache.flink.client.deployment.application.ApplicationDispatcherBootstrap.runApplicationEntryPoint(ApplicationDispatcherBootstrap.java:261) ~[flink-dist_2.11-1.14.6.jar:1.14.6]
	... 12 more
2023-08-25 11:42:33,169 INFO  org.apache.flink.runtime.entrypoint.ClusterEntrypoint        [] - Shutting YarnApplicationClusterEntryPoint down with application status UNKNOWN. Diagnostics Cluster entrypoint has been closed externally..
```


### Screenshots

<img width="1570" alt="image" src="https://github.com/apache/incubator-streampark/assets/38800374/f2b2776f-5f6f-47be-9615-b8b82e6f0122">


### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!(您是否要贡献这个PR?)

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
+++ b/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
@@ -471,6 +471,15 @@ trait FlinkClientTrait extends Logger {
       }
 
     }
+
+    // execution.runtime-mode
+    if (submitRequest.properties.nonEmpty) {
+      if (submitRequest.properties.containsKey(ExecutionOptions.RUNTIME_MODE.key())) {
+        programArgs += s"--${ExecutionOptions.RUNTIME_MODE.key()}"
+        programArgs += submitRequest.properties.get(ExecutionOptions.RUNTIME_MODE.key())
+      }
+    }
+
     programArgs.toList.asJava
   }
 

--- a/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
+++ b/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
@@ -476,7 +476,7 @@ trait FlinkClientTrait extends Logger {
     if (submitRequest.properties.nonEmpty) {
       if (submitRequest.properties.containsKey(ExecutionOptions.RUNTIME_MODE.key())) {
         programArgs += s"--${ExecutionOptions.RUNTIME_MODE.key()}"
-        programArgs += submitRequest.properties.get(ExecutionOptions.RUNTIME_MODE.key())
+        programArgs += submitRequest.properties.get(ExecutionOptions.RUNTIME_MODE.key()).toString
       }
     }
 

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/EnhancerImplicit.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/EnhancerImplicit.scala
@@ -21,6 +21,9 @@ import org.apache.streampark.common.conf.ConfigConst.{KEY_APP_NAME, KEY_FLINK_AP
 import org.apache.streampark.common.util.DeflaterUtils
 
 import org.apache.flink.api.java.utils.ParameterTool
+import org.apache.flink.configuration.PipelineOptions
+import org.apache.flink.table.api.TableEnvironment
+import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
 
 import scala.util.Try
 
@@ -28,7 +31,7 @@ object EnhancerImplicit {
 
   implicit class EnhanceParameterTool(parameterTool: ParameterTool) {
 
-    def getAppName(name: String = null, required: Boolean = false): String = {
+    private[flink] def getAppName(name: String = null, required: Boolean = false): String = {
       val appName = name match {
         case null =>
           Try(DeflaterUtils.unzipString(parameterTool.get(KEY_APP_NAME(), null)))
@@ -41,6 +44,29 @@ object EnhancerImplicit {
       appName
     }
 
+  }
+
+  implicit class EnhanceTableEnvironment(env: TableEnvironment) {
+
+    private[flink] def setAppName(implicit parameter: ParameterTool): TableEnvironment = {
+      val appName = parameter.getAppName()
+      if (appName != null) {
+        env.getConfig.getConfiguration.setString(PipelineOptions.NAME, appName)
+      }
+      env
+    }
+
+  }
+
+  implicit class EnhanceStreamExecutionEnvironment(env: StreamTableEnvironment) {
+
+    private[flink] def setAppName(implicit parameter: ParameterTool): StreamTableEnvironment = {
+      val appName = parameter.getAppName()
+      if (appName != null) {
+        env.getConfig.getConfiguration.setString(PipelineOptions.NAME, appName)
+      }
+      env
+    }
   }
 
 }

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/FlinkStreamingInitializer.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/FlinkStreamingInitializer.scala
@@ -34,33 +34,17 @@ import java.io.File
 
 private[flink] object FlinkStreamingInitializer {
 
-  private[this] var flinkInitializer: FlinkStreamingInitializer = _
-
   def initialize(args: Array[String], config: (StreamExecutionEnvironment, ParameterTool) => Unit)
       : (ParameterTool, StreamExecutionEnvironment) = {
-    if (flinkInitializer == null) {
-      this.synchronized {
-        if (flinkInitializer == null) {
-          flinkInitializer = new FlinkStreamingInitializer(args, ApiType.scala)
-          flinkInitializer.streamEnvConfFunc = config
-          flinkInitializer.initEnvironment()
-        }
-      }
-    }
-    (flinkInitializer.configuration.parameter, flinkInitializer.streamEnvironment)
+    val flinkInitializer = new FlinkStreamingInitializer(args, ApiType.scala)
+    flinkInitializer.streamEnvConfFunc = config
+    (flinkInitializer.configuration.parameter, flinkInitializer.streamEnv)
   }
 
   def initialize(args: StreamEnvConfig): (ParameterTool, StreamExecutionEnvironment) = {
-    if (flinkInitializer == null) {
-      this.synchronized {
-        if (flinkInitializer == null) {
-          flinkInitializer = new FlinkStreamingInitializer(args.args, ApiType.java)
-          flinkInitializer.javaStreamEnvConfFunc = args.conf
-          flinkInitializer.initEnvironment()
-        }
-      }
-    }
-    (flinkInitializer.configuration.parameter, flinkInitializer.streamEnvironment)
+    val flinkInitializer = new FlinkStreamingInitializer(args.args, ApiType.java)
+    flinkInitializer.javaStreamEnvConfFunc = args.conf
+    (flinkInitializer.configuration.parameter, flinkInitializer.streamEnv)
   }
 }
 
@@ -75,7 +59,22 @@ private[flink] class FlinkStreamingInitializer(args: Array[String], apiType: Api
 
   var javaTableEnvConfFunc: TableEnvConfigFunction = _
 
-  private[this] var localStreamEnv: StreamExecutionEnvironment = _
+  implicit private[flink] val parameter: ParameterTool = configuration.parameter
+
+  lazy val streamEnv: StreamExecutionEnvironment = {
+    val env = new StreamExecutionEnvironment(
+      JavaStreamEnv.getExecutionEnvironment(configuration.envConfig))
+
+    apiType match {
+      case ApiType.java if javaStreamEnvConfFunc != null =>
+        javaStreamEnvConfFunc.configuration(env.getJavaEnv, configuration.parameter)
+      case ApiType.scala if streamEnvConfFunc != null =>
+        streamEnvConfFunc(env, configuration.parameter)
+      case _ =>
+    }
+    env.getConfig.setGlobalJobParameters(configuration.parameter)
+    env
+  }
 
   lazy val configuration: FlinkConfiguration = initParameter()
 
@@ -145,31 +144,6 @@ private[flink] class FlinkStreamingInitializer(args: Array[String], apiType: Api
           map += x._1.drop(prefix.length) -> x._2
         })
     map
-  }
-
-  def streamEnvironment: StreamExecutionEnvironment = {
-    if (localStreamEnv == null) {
-      this.synchronized {
-        if (localStreamEnv == null) {
-          initEnvironment()
-        }
-      }
-    }
-    localStreamEnv
-  }
-
-  def initEnvironment(): Unit = {
-    localStreamEnv = new StreamExecutionEnvironment(
-      JavaStreamEnv.getExecutionEnvironment(configuration.envConfig))
-
-    apiType match {
-      case ApiType.java if javaStreamEnvConfFunc != null =>
-        javaStreamEnvConfFunc.configuration(localStreamEnv.getJavaEnv, configuration.parameter)
-      case ApiType.scala if streamEnvConfFunc != null =>
-        streamEnvConfFunc(localStreamEnv, configuration.parameter)
-      case _ =>
-    }
-    localStreamEnv.getConfig.setGlobalJobParameters(configuration.parameter)
   }
 
 }

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/FlinkTableInitializer.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/FlinkTableInitializer.scala
@@ -17,15 +17,14 @@
 package org.apache.streampark.flink.core
 
 import org.apache.streampark.common.conf.ConfigConst._
-import org.apache.streampark.common.enums.{ApiType, PlannerType, TableMode}
+import org.apache.streampark.common.enums.{ApiType, PlannerType}
 import org.apache.streampark.common.enums.ApiType.ApiType
-import org.apache.streampark.common.enums.TableMode.TableMode
 import org.apache.streampark.common.util.{DeflaterUtils, PropertiesUtils}
 import org.apache.streampark.flink.core.EnhancerImplicit._
 import org.apache.streampark.flink.core.conf.FlinkConfiguration
 
 import org.apache.flink.api.java.utils.ParameterTool
-import org.apache.flink.configuration.{Configuration, PipelineOptions}
+import org.apache.flink.configuration.Configuration
 import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
 import org.apache.flink.table.api.{EnvironmentSettings, TableConfig, TableEnvironment}
 import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
@@ -38,34 +37,18 @@ import scala.util.{Failure, Success, Try}
 
 private[flink] object FlinkTableInitializer {
 
-  private[this] var flinkInitializer: FlinkTableInitializer = _
-
   def initialize(
       args: Array[String],
       config: (TableConfig, ParameterTool) => Unit): (ParameterTool, TableEnvironment) = {
-    if (flinkInitializer == null) {
-      this.synchronized {
-        if (flinkInitializer == null) {
-          flinkInitializer = new FlinkTableInitializer(args, ApiType.scala)
-          flinkInitializer.tableConfFunc = config
-          flinkInitializer.initEnvironment(TableMode.batch)
-        }
-      }
-    }
-    (flinkInitializer.configuration.parameter, flinkInitializer.tableEnvironment)
+    val flinkInitializer = new FlinkTableInitializer(args, ApiType.scala)
+    flinkInitializer.tableConfFunc = config
+    (flinkInitializer.configuration.parameter, flinkInitializer.tableEnv)
   }
 
   def initialize(args: TableEnvConfig): (ParameterTool, TableEnvironment) = {
-    if (flinkInitializer == null) {
-      this.synchronized {
-        if (flinkInitializer == null) {
-          flinkInitializer = new FlinkTableInitializer(args.args, ApiType.java)
-          flinkInitializer.javaTableEnvConfFunc = args.conf
-          flinkInitializer.initEnvironment(TableMode.batch)
-        }
-      }
-    }
-    (flinkInitializer.configuration.parameter, flinkInitializer.tableEnvironment)
+    val flinkInitializer = new FlinkTableInitializer(args.args, ApiType.java)
+    flinkInitializer.javaTableEnvConfFunc = args.conf
+    (flinkInitializer.configuration.parameter, flinkInitializer.tableEnv)
   }
 
   def initialize(
@@ -73,38 +56,25 @@ private[flink] object FlinkTableInitializer {
       configStream: (StreamExecutionEnvironment, ParameterTool) => Unit,
       configTable: (TableConfig, ParameterTool) => Unit)
       : (ParameterTool, StreamExecutionEnvironment, StreamTableEnvironment) = {
-    if (flinkInitializer == null) {
-      this.synchronized {
-        if (flinkInitializer == null) {
-          flinkInitializer = new FlinkTableInitializer(args, ApiType.scala)
-          flinkInitializer.streamEnvConfFunc = configStream
-          flinkInitializer.tableConfFunc = configTable
-          flinkInitializer.initEnvironment(TableMode.streaming)
-        }
-      }
-    }
+
+    val flinkInitializer = new FlinkTableInitializer(args, ApiType.scala)
+    flinkInitializer.streamEnvConfFunc = configStream
+    flinkInitializer.tableConfFunc = configTable
     (
       flinkInitializer.configuration.parameter,
-      flinkInitializer.streamEnvironment,
-      flinkInitializer.streamTableEnvironment)
+      flinkInitializer.streamEnv,
+      flinkInitializer.streamTableEnv)
   }
 
   def initialize(args: StreamTableEnvConfig)
       : (ParameterTool, StreamExecutionEnvironment, StreamTableEnvironment) = {
-    if (flinkInitializer == null) {
-      this.synchronized {
-        if (flinkInitializer == null) {
-          flinkInitializer = new FlinkTableInitializer(args.args, ApiType.java)
-          flinkInitializer.javaStreamEnvConfFunc = args.streamConfig
-          flinkInitializer.javaTableEnvConfFunc = args.tableConfig
-          flinkInitializer.initEnvironment(TableMode.streaming)
-        }
-      }
-    }
+    val flinkInitializer = new FlinkTableInitializer(args.args, ApiType.java)
+    flinkInitializer.javaStreamEnvConfFunc = args.streamConfig
+    flinkInitializer.javaTableEnvConfFunc = args.tableConfig
     (
       flinkInitializer.configuration.parameter,
-      flinkInitializer.streamEnvironment,
-      flinkInitializer.streamTableEnvironment)
+      flinkInitializer.streamEnv,
+      flinkInitializer.streamTableEnv)
   }
 
 }
@@ -112,30 +82,95 @@ private[flink] object FlinkTableInitializer {
 private[flink] class FlinkTableInitializer(args: Array[String], apiType: ApiType)
   extends FlinkStreamingInitializer(args, apiType) {
 
-  private[this] var localStreamTableEnv: StreamTableEnvironment = _
+  private[this] lazy val envSettings = {
 
-  private[this] var localTableEnv: TableEnvironment = _
+    val builder = EnvironmentSettings.newInstance()
 
-  def streamTableEnvironment: StreamTableEnvironment = {
-    if (localStreamTableEnv == null) {
-      this.synchronized {
-        if (localStreamTableEnv == null) {
-          initEnvironment(TableMode.streaming)
+    Try(PlannerType.withName(parameter.get(KEY_FLINK_TABLE_PLANNER)))
+      .getOrElse(PlannerType.blink) match {
+      case PlannerType.blink =>
+        val useBlinkPlanner =
+          Try(builder.getClass.getDeclaredMethod("useBlinkPlanner")).getOrElse(null)
+        if (useBlinkPlanner == null) {
+          logWarn("useBlinkPlanner deprecated")
+        } else {
+          useBlinkPlanner.setAccessible(true)
+          useBlinkPlanner.invoke(builder)
+          logInfo("blinkPlanner will be use.")
         }
-      }
+      case PlannerType.old =>
+        val useOldPlanner = Try(builder.getClass.getDeclaredMethod("useOldPlanner")).getOrElse(null)
+        if (useOldPlanner == null) {
+          logWarn("useOldPlanner deprecated")
+        } else {
+          useOldPlanner.setAccessible(true)
+          useOldPlanner.invoke(builder)
+          logInfo("useOldPlanner will be use.")
+        }
+      case PlannerType.any =>
+        val useAnyPlanner = Try(builder.getClass.getDeclaredMethod("useAnyPlanner")).getOrElse(null)
+        if (useAnyPlanner == null) {
+          logWarn("useAnyPlanner deprecated")
+        } else {
+          logInfo("useAnyPlanner will be use.")
+          useAnyPlanner.setAccessible(true)
+          useAnyPlanner.invoke(builder)
+        }
     }
-    localStreamTableEnv
+
+    val buildWith =
+      (parameter.get(KEY_FLINK_TABLE_CATALOG), parameter.get(KEY_FLINK_TABLE_DATABASE))
+    buildWith match {
+      case (x: String, y: String) if x != null && y != null =>
+        logInfo(s"with built in catalog: $x")
+        logInfo(s"with built in database: $y")
+        builder.withBuiltInCatalogName(x)
+        builder.withBuiltInDatabaseName(y)
+      case (x: String, _) if x != null =>
+        logInfo(s"with built in catalog: $x")
+        builder.withBuiltInCatalogName(x)
+      case (_, y: String) if y != null =>
+        logInfo(s"with built in database: $y")
+        builder.withBuiltInDatabaseName(y)
+      case _ =>
+    }
+    builder
   }
 
-  def tableEnvironment: TableEnvironment = {
-    if (localTableEnv == null) {
-      this.synchronized {
-        if (localTableEnv == null) {
-          initEnvironment(TableMode.batch)
-        }
-      }
+  lazy val tableEnv: TableEnvironment = {
+    logInfo(s"job working in batch mode")
+    envSettings.inBatchMode()
+    val tableEnv = TableEnvironment.create(envSettings.build()).setAppName
+    apiType match {
+      case ApiType.java if javaTableEnvConfFunc != null =>
+        javaTableEnvConfFunc.configuration(tableEnv.getConfig, parameter)
+      case ApiType.scala if tableConfFunc != null =>
+        tableConfFunc(tableEnv.getConfig, parameter)
+      case _ =>
     }
-    localTableEnv
+    tableEnv
+  }
+
+  lazy val streamTableEnv: StreamTableEnvironment = {
+    logInfo(s"components should work in streaming mode")
+    envSettings.inStreamingMode()
+    val setting = envSettings.build()
+
+    if (streamEnvConfFunc != null) {
+      streamEnvConfFunc(streamEnv, parameter)
+    }
+    if (javaStreamEnvConfFunc != null) {
+      javaStreamEnvConfFunc.configuration(streamEnv.getJavaEnv, parameter)
+    }
+    val streamTableEnv = StreamTableEnvironment.create(streamEnv, setting).setAppName
+    apiType match {
+      case ApiType.java if javaTableEnvConfFunc != null =>
+        javaTableEnvConfFunc.configuration(streamTableEnv.getConfig, parameter)
+      case ApiType.scala if tableConfFunc != null =>
+        tableConfFunc(streamTableEnv.getConfig, parameter)
+      case _ =>
+    }
+    streamTableEnv
   }
 
   /** In case of table SQL, the parameter conf is not required, it depends on the developer. */
@@ -197,110 +232,6 @@ private[flink] class FlinkTableInitializer(args: Array[String], apiType: ApiType
                 new IllegalArgumentException(s"[StreamPark] init sql error.$e")
                 configuration
             }
-        }
-    }
-  }
-
-  def initEnvironment(tableMode: TableMode): Unit = {
-    val builder = EnvironmentSettings.newInstance()
-    val parameter = configuration.parameter
-    Try(PlannerType.withName(parameter.get(KEY_FLINK_TABLE_PLANNER)))
-      .getOrElse(PlannerType.blink) match {
-      case PlannerType.blink =>
-        val useBlinkPlanner =
-          Try(builder.getClass.getDeclaredMethod("useBlinkPlanner")).getOrElse(null)
-        if (useBlinkPlanner == null) {
-          logWarn("useBlinkPlanner deprecated")
-        } else {
-          useBlinkPlanner.setAccessible(true)
-          useBlinkPlanner.invoke(builder)
-          logInfo("blinkPlanner will be use.")
-        }
-      case PlannerType.old =>
-        val useOldPlanner = Try(builder.getClass.getDeclaredMethod("useOldPlanner")).getOrElse(null)
-        if (useOldPlanner == null) {
-          logWarn("useOldPlanner deprecated")
-        } else {
-          useOldPlanner.setAccessible(true)
-          useOldPlanner.invoke(builder)
-          logInfo("useOldPlanner will be use.")
-        }
-      case PlannerType.any =>
-        val useAnyPlanner = Try(builder.getClass.getDeclaredMethod("useAnyPlanner")).getOrElse(null)
-        if (useAnyPlanner == null) {
-          logWarn("useAnyPlanner deprecated")
-        } else {
-          logInfo("useAnyPlanner will be use.")
-          useAnyPlanner.setAccessible(true)
-          useAnyPlanner.invoke(builder)
-        }
-    }
-
-    val mode = Try(TableMode.withName(parameter.get(KEY_FLINK_TABLE_MODE))).getOrElse(tableMode)
-    mode match {
-      case TableMode.batch =>
-        logInfo(s"components should work in $tableMode mode")
-        builder.inBatchMode()
-      case TableMode.streaming =>
-        logInfo(s"components should work in $tableMode mode")
-        builder.inStreamingMode()
-    }
-
-    val buildWith =
-      (parameter.get(KEY_FLINK_TABLE_CATALOG), parameter.get(KEY_FLINK_TABLE_DATABASE))
-    buildWith match {
-      case (x: String, y: String) if x != null && y != null =>
-        logInfo(s"with built in catalog: $x")
-        logInfo(s"with built in database: $y")
-        builder.withBuiltInCatalogName(x)
-        builder.withBuiltInDatabaseName(y)
-      case (x: String, _) if x != null =>
-        logInfo(s"with built in catalog: $x")
-        builder.withBuiltInCatalogName(x)
-      case (_, y: String) if y != null =>
-        logInfo(s"with built in database: $y")
-        builder.withBuiltInDatabaseName(y)
-      case _ =>
-    }
-    val setting = builder.build()
-    tableMode match {
-      case TableMode.batch => localTableEnv = TableEnvironment.create(setting)
-      case TableMode.streaming =>
-        initEnvironment()
-        if (streamEnvConfFunc != null) {
-          streamEnvConfFunc(streamEnvironment, parameter)
-        }
-        if (javaStreamEnvConfFunc != null) {
-          javaStreamEnvConfFunc.configuration(streamEnvironment.getJavaEnv, parameter)
-        }
-        localStreamTableEnv = StreamTableEnvironment.create(streamEnvironment, setting)
-    }
-    val appName = parameter.getAppName()
-    if (appName != null) {
-      tableMode match {
-        case TableMode.batch =>
-          localTableEnv.getConfig.getConfiguration.setString(PipelineOptions.NAME, appName)
-        case TableMode.streaming =>
-          localStreamTableEnv.getConfig.getConfiguration.setString(PipelineOptions.NAME, appName)
-      }
-    }
-
-    apiType match {
-      case ApiType.java =>
-        if (javaTableEnvConfFunc != null) {
-          tableMode match {
-            case TableMode.batch =>
-              javaTableEnvConfFunc.configuration(localTableEnv.getConfig, parameter)
-            case TableMode.streaming =>
-              javaTableEnvConfFunc.configuration(localStreamTableEnv.getConfig, parameter)
-          }
-        }
-      case ApiType.scala =>
-        if (tableConfFunc != null) {
-          tableMode match {
-            case TableMode.batch => tableConfFunc(localTableEnv.getConfig, parameter)
-            case TableMode.streaming => tableConfFunc(localStreamTableEnv.getConfig, parameter)
-          }
         }
     }
   }

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.12/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.12/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
@@ -47,7 +47,7 @@ class TableContext(override val parameter: ParameterTool, private val tableEnv: 
 
   override def execute(jobName: String): JobExecutionResult = {
     printLogo(s"FlinkTable $jobName Starting...")
-    tableEnv.execute(jobName)
+    null
   }
 
   @deprecated override def fromTableSource(source: TableSource[_]): Table =

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.13/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.13/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
@@ -52,7 +52,7 @@ class TableContext(override val parameter: ParameterTool, private val tableEnv: 
 
   override def execute(jobName: String): JobExecutionResult = {
     printLogo(s"FlinkTable $jobName Starting...")
-    tableEnv.execute(jobName)
+    null
   }
 
   @deprecated override def fromTableSource(source: TableSource[_]): Table =

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.14/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.14/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
@@ -65,7 +65,7 @@ class TableContext(override val parameter: ParameterTool, private val tableEnv: 
 
   @deprecated override def execute(jobName: String): JobExecutionResult = {
     printLogo(s"FlinkTable $jobName Starting...")
-    tableEnv.execute(jobName)
+    null
   }
 
   @deprecated override def fromTableSource(source: TableSource[_]): Table =

--- a/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.16/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
+++ b/streampark-flink/streampark-flink-shims/streampark-flink-shims_flink-1.16/src/main/scala/org/apache/streampark/flink/core/TableContext.scala
@@ -18,7 +18,7 @@
 package org.apache.streampark.flink.core
 
 import org.apache.flink.api.java.utils.ParameterTool
-import org.apache.flink.table.api.{CompiledPlan, PlanReference, Table, TableDescriptor, TableEnvironment, TableResult}
+import org.apache.flink.table.api.{CompiledPlan, PlanReference, Table, TableDescriptor, TableEnvironment}
 import org.apache.flink.table.module.ModuleEntry
 
 class TableContext(override val parameter: ParameterTool, private val tableEnv: TableEnvironment)

--- a/streampark-flink/streampark-flink-sqlclient/src/main/scala/org/apache/streampark/flink/cli/SqlClient.scala
+++ b/streampark-flink/streampark-flink-sqlclient/src/main/scala/org/apache/streampark/flink/cli/SqlClient.scala
@@ -26,30 +26,42 @@ import org.apache.flink.api.java.utils.ParameterTool
 import org.apache.flink.configuration.ExecutionOptions
 
 import scala.language.implicitConversions
-import scala.util.Try
+import scala.util.{Failure, Success, Try}
 
 object SqlClient extends App {
 
-  val parameterTool = ParameterTool.fromArgs(args)
+  private[this] val parameterTool = ParameterTool.fromArgs(args)
 
-  val flinkSql = {
+  private[this] val flinkSql = {
     val sql = parameterTool.get(KEY_FLINK_SQL())
     require(sql != null && sql.trim.nonEmpty, "Usage: flink sql cannot be null")
-    Try(DeflaterUtils.unzipString(sql)).getOrElse(
-      throw new IllegalArgumentException("Usage: flink sql is invalid or null, please check"))
+    Try(DeflaterUtils.unzipString(sql)) match {
+      case Success(value) => value
+      case Failure(_) =>
+        throw new IllegalArgumentException("Usage: flink sql is invalid or null, please check")
+    }
   }
 
-  val sets = SqlCommandParser.parseSQL(flinkSql).filter(_.command == SqlCommand.SET)
+  private[this] val sets = SqlCommandParser.parseSQL(flinkSql).filter(_.command == SqlCommand.SET)
 
-  val mode = sets.find(_.operands.head == ExecutionOptions.RUNTIME_MODE.key()) match {
-    case Some(e) => e.operands(1)
+  private[this] val defaultMode = "streaming"
+
+  private[this] val mode = sets.find(_.operands.head == ExecutionOptions.RUNTIME_MODE.key()) match {
+    case Some(e) =>
+      // 1) flink sql execution.runtime-mode has highest priority
+      e.operands(1)
     case None =>
-      val appConf = parameterTool.get(KEY_APP_CONF(), null)
-      val defaultMode = "streaming"
-      if (appConf == null) defaultMode
-      else {
-        val parameter = PropertiesUtils.fromYamlText(DeflaterUtils.unzipString(appConf.drop(7)))
-        parameter.getOrElse(KEY_FLINK_TABLE_MODE, defaultMode)
+      // 2) dynamic properties execution.runtime-mode
+      parameterTool.get(ExecutionOptions.RUNTIME_MODE.key(), null) match {
+        case null =>
+          parameterTool.get(KEY_APP_CONF(), null) match {
+            case null => defaultMode
+            case f =>
+              val parameter = PropertiesUtils.fromYamlText(DeflaterUtils.unzipString(f.drop(7)))
+              // 3) application conf execution.runtime-mode
+              parameter.getOrElse(KEY_FLINK_TABLE_MODE, defaultMode)
+          }
+        case m => m
       }
   }
 

